### ISSUE TITLE
learning with discriminator.json converges too slowly

### ISSUE THREAD

**Issue by perprit** [naver/gdc#2]
Hi, thanks for the great work.

I'm trying to reproduce the experiment on pointwise constraints.
However, it seems the training converges too slowly specifically on `discriminator.json`, in contrast to other configurations.
After 7 days of training on 2 V100 gpus, I found the `Eval/b(x)_mean` is still fluctuating around 0.25~0.28 (compared to the result reported on paper). It seems `Eval/KL(p || pi)` is still decreasing, so I believe it's not converged yet.

Are there any differences between the configuration on this repo and the one the model that gives the result on the paper is trained with?
Or, could you give any recommendation to boost the training?

Thanks in advance!

**Comment by mukhal** [comment]
Hi @perprit,

Thank you for your interest in our work. 

The code in the repo is the exact same code used to produce the results in the paper. We have found out that batch size plays a major role in the convergence speed. In our experiments, we used a batch size of 2048. So, one suggestion for you is to use a larger batch size if possible. On a side note and as we report in the paper, some types of constraints are harder to satisfy than others, and in these cases, we have found `Eval/b(x)_mean` to plateau after training for some time. 

Edit: we noticed that the learning rate used in `discriminator.json` is lower than the one we used in the paper. So as Hady suggested, a higher learning rate will likely boost convergence speed.

**Comment by hadyelsahar** [comment]
Hi @perprit  thanks for your interest in our work. This depends on which classifier you are using is that the `sentiment` one with class index `[0,2]`? or have you changed that? If you could put a screenshot with your KL(p,pi) and b(x) curves we might be able to help you more. 

> Or, could you give any recommendation to boost the training?

Overall, you could try increasing the learning rate this usually helps b(x) to increase faster but sometimes introduce noise and `KL(p,pi)` starts to diverge instead of decreasing. 
I would try higher values such as `0.1 e-5` and `0.5 e-5` if you are using a large batch size `2048`your training should go faster. But If KL(p,pi) fluctuates or doesn't decrease anymore then try lowering the learning rate until you reach a good value.

**Comment by perprit** [comment]
Hi, thanks for the quick replies.
I trained the model with `discriminator.json` without any changes. As you mentioned, I also found the learning rate is set `0.003e-5`, which is 1000 times lower compared to `1.41e-5`. Hence I trained the model with the same config while only modifying the learning rate to `1.41e-5`. Now I found the training converges faster, but also `KL(p, pi)` seems to fluctuate as @hadyelsahar said. Anyway, I guess it would converge to the result on the paper soon. Thank you all!


```
lr: 0.003e-5
"Eval/b(x)_mean": 0.12646484375, "Eval/KL(p || pi)": 2.131262879177428,
"Eval/b(x)_mean": 0.13623046875, "Eval/KL(p || pi)": 2.1225270825649365,
"Eval/b(x)_mean": 0.13623046875, "Eval/KL(p || pi)": 2.1070878404059497,
"Eval/b(x)_mean": 0.1474609375, "Eval/KL(p || pi)": 2.091780795628608,
"Eval/b(x)_mean": 0.142578125, "Eval/KL(p || pi)": 2.0914833170167864,
"Eval/b(x)_mean": 0.13671875, "Eval/KL(p || pi)": 2.0621614009246962,
"Eval/b(x)_mean": 0.12646484375, "Eval/KL(p || pi)": 2.1357987336253332,
"Eval/b(x)_mean": 0.138671875, "Eval/KL(p || pi)": 2.128864284672665,
"Eval/b(x)_mean": 0.14501953125, "Eval/KL(p || pi)": 2.1231345740092644,
"Eval/b(x)_mean": 0.1455078125, "Eval/KL(p || pi)": 2.0991868776093936,

lr: 1.41e-5
"Eval/b(x)_mean": 0.119140625, "Eval/KL(p || pi)": 3.22326802464684,
"Eval/b(x)_mean": 0.18017578125, "Eval/KL(p || pi)": 3.641144157105698,
"Eval/b(x)_mean": 0.2099609375, "Eval/KL(p || pi)": 2.5374318871919104,
"Eval/b(x)_mean": 0.23974609375, "Eval/KL(p || pi)": 2.3262977679572736,
"Eval/b(x)_mean": 0.28173828125, "Eval/KL(p || pi)": 2.7291163615729994,
"Eval/b(x)_mean": 0.28955078125, "Eval/KL(p || pi)": 2.6127682956186273,
"Eval/b(x)_mean": 0.3095703125, "Eval/KL(p || pi)": 2.4726754928901515,
"Eval/b(x)_mean": 0.32421875, "Eval/KL(p || pi)": 2.563776352539344,
"Eval/b(x)_mean": 0.3095703125, "Eval/KL(p || pi)": 2.6010831468831723,
"Eval/b(x)_mean": 0.3173828125, "Eval/KL(p || pi)": 2.6151738753663434,
"Eval/b(x)_mean": 0.35302734375, "Eval/KL(p || pi)": 2.3390441769125534,
"Eval/b(x)_mean": 0.3486328125, "Eval/KL(p || pi)": 2.9610756016326034,
```

**Comment by hadyelsahar** [comment]
great it worked, closing this for now, feel free to reopen if other issues arrive.
